core/src/main/java/com/linecorp/armeria/internal/common/stream/ByteBufsInputStream.java

```diff
@@ -150,6 +150,7 @@ public void setEos() {
     public void add(ByteBuf buffer) {
         requireNonNull(buffer, "buffer");
         if (eos.get()) {
+            buffer.release();
             throw new IllegalStateException("Already closed");
         }
         if (!buffers.add(buffer)) {
```

Comment:
> Thanks for the fix. 😉
How about also raising an exception if an empty byte buf is added?
... or Probably better just silently releasing? https://github.com/line/armeria/pull/4271#issuecomment-1148133323

```java
ByteBufsInputStream stream = new ByteBufsInputStream();
ByteBuf buffer = Unpooled.buffer(); // empty at first
stream.add(buffer);

buffer.writeByte(1); // write 1
stream.read() // return 1
```
Hmm..  maybe `ByteBufsInputStream` allow to add `empty` ByteBuf,
So I'm not sure we should `release` or `raising` an exception if an `empty` ByteBuf is added. WDYT? 🤔
cc. @minwoox 

https://github.com/line/armeria/blob/7561be45824b691a32e486d8d8a98c3ad930c32b/core/src/main/java/com/linecorp/armeria/internal/common/stream/ByteBufsInputStream.java#L224-L226
(`ByteBufsInputStream` already handles `empty` ByteBuf inside)

Reply:
> Hmm.. maybe ByteBufsInputStream allow to add empty ByteBuf

So my opinion is just keeping current implementation! (allow to add `empty` ByteBuf on `ByteBufsInputStream`)

Reply:
Let's say that the following `ByteBufs` are added consecutively:
```
ByteBuf(3 bytes) - ByteBuf(0 byte) - ByteBuf(4 bytes) 
```
Even though we have more bytes to read, we can't get the last `ByteBuf` that has 4 bytes in it.
To prevent this, I think we have two possible solutions:
- Return `0` instead of `-1` when `buffer.isReadable()` returns `false` so that we can keep continue to read.
- Do not allow the empty buffer to get added.

Because we add `EMPTY_BUFFER` in the `setEos` method to tell the end of the bytes, if we go to the first solution, we need more work to do. (e.g. adding the special empty buffer instead of `Unpooled.EMPTY_BUFFER` to indicate eos.)
So I prefer the second one which silently releases it and does not add the queue as @trustin suggested.
What do you think? @injae-kim

Reply:
> ByteBuf(3 bytes) - ByteBuf(0 byte) - ByteBuf(4 bytes) 

oh.. I understood your explanation. I didn't know this case!

> I prefer the second one which silently releases it and does not add the queue as @trustin suggested.

I'll update PR in this way! thanks 👍

Reply:
https://github.com/line/armeria/blob/7561be45824b691a32e486d8d8a98c3ad930c32b/core/src/main/java/com/linecorp/armeria/internal/common/stream/ByteBufsInputStream.java#L208-L211

> ByteBuf(3 bytes) - ByteBuf(0 byte) - ByteBuf(4 bytes) 

Oh.. after some checking, I found that there's already handling logic for empty bytebuf,
so I think above case is not occurred (`ByteBuf(0 byte)` will be passed during read())

Anyway, I address @minwoox comment and enhance unit test for adding empty bytebuf case 🙇

Reply:
Ah, I didn't know that we're already handling the case. 😅  thanks.